README.md
# NAME

XML::Invisible - transform "invisible XML" documents into XML using a grammar

# PROJECT STATUS

| OS      |  Build status |
|:-------:|--------------:|
| Linux   | [![Build Status](https://travis-ci.org/mohawk2/xml-invisible.svg?branch=master)](https://travis-ci.org/mohawk2/xml-invisible) |

[![CPAN version](https://badge.fury.io/pl/XML-Invisible.svg)](https://metacpan.org/pod/XML::Invisible) [![Coverage Status](https://coveralls.io/repos/github/mohawk2/xml-invisible/badge.svg?branch=master)](https://coveralls.io/github/mohawk2/xml-invisible?branch=master)

# SYNOPSIS

    use XML::Invisible qw(make_parser);
    my $transformer = make_parser(from_file($ixmlspec));
    my $ast = $transformer->(from_file($ixml_input));

    use XML::Invisible qw(make_parser ast2xml);
    my $transformer = make_parser(from_file($ixmlspec));
    my $xmldoc = ast2xml($transformer->(from_file($ixml_input)));
    to_file($outputfile, $xmldoc->toStringC14N(1));

    # or, with conventional pre-compiled Pegex grammar
    my $parser = Pegex::Parser->new(
      grammar => My::Thing::Grammar->new,
      receiver => XML::Invisible::Receiver->new,
    );
    my $got = $parser->parse($input);

    # from command line
    cpanm XML::Invisible XML::Twig
    perl -MXML::Invisible=make_parser,ast2xml -e \
      'print ast2xml(make_parser(join "", <>)->("(a+b)"))->toStringC14N(1)' \
      examples/arith-grammar.ixml | xml_pp

# DESCRIPTION

An implementation of Steven Pemberton's Invisible XML concept, using
[Pegex](https://metacpan.org/pod/Pegex). Supply it with your grammar, in Pegex format (slightly
different from Steven's specification due to differences between his
notation and Pegex's), it returns you a function, which you can call to
transform "invisible XML" documents into actual XML.

This is largely a Pegex "receiver" class that exploits the `+` and `-`
syntax in rules in slightly unintended ways, and a wrapper to make
this operate.

# GRAMMAR SYNTAX

See [Pegex::Syntax](https://metacpan.org/pod/Pegex::Syntax). Generally, all rules will result in an XML
element. All terminals will need to capture with `()` (see example
below).

However, if you specify a dependent token with `+` it will
instead become an attribute (equivalent of Steven's `@`). If `-`,
this will "flatten" (equivalent of Steven's `-`) - the children will
be included without making an element of that node. Since in Pegex any
element can be skipped entirely with `.`, you can use that instead of
`-` to omit terminals.

E.g.

    expr: +open -arith +close
    open: /( LPAREN )/
    close: /( RPAREN )/
    arith: left -op right
    left: +name
    right: -name
    name: /(a)/ | /(b)/
    op: +sign
    sign: /( PLUS )/

When given `(a+b)` yields:

    <expr open="(" sign="+" close=")">
      <left name="a"/>
      <right>b</right>
    </expr>

# FUNCTIONS

## make\_parser

Exportable. Returns a function that when called with an "invisible XML"
document, it will return an abstract syntax tree (AST), of the general form:

    {
      nodename => 'expr',
      attributes => { open => '(', sign => '+', close => ')' },
      children => [
        { nodename => 'left', attributes => { name => 'a' } },
        { nodename => 'right', children => [ 'b' ] },
      ],
    }

Arguments:

- an "invisible XML" grammar specification, in Pegex format

See [XML::Invisible::Receiver](https://metacpan.org/pod/XML::Invisible::Receiver) for more.

## ast2xml

Exportable. Turns an AST, as output by ["make\_parser"](#make_parser),
from [XML::Invisible::Receiver](https://metacpan.org/pod/XML::Invisible::Receiver) into an object of class
[XML::LibXML::Document](https://metacpan.org/pod/XML::LibXML::Document). Needs [XML::LibXML](https://metacpan.org/pod/XML::LibXML) installed, which as of
version 0.05 of this module is only a suggested dependency, not required.

Arguments:

- an AST from [XML::Invisible::Receiver](https://metacpan.org/pod/XML::Invisible::Receiver)

# DEBUGGING

To debug, set environment variable `XML_INVISIBLE_DEBUG` to a true value.

# SEE ALSO

[Pegex](https://metacpan.org/pod/Pegex)

[https://homepages.cwi.nl/~steven/ixml/](https://homepages.cwi.nl/~steven/ixml/) - Steven Pemberton's Invisible XML page

# AUTHOR

Ed J, `<etj at cpan.org>`

# BUGS

Please report any bugs or feature requests on
[https://github.com/mohawk2/xml-invisible/issues](https://github.com/mohawk2/xml-invisible/issues).

Or, if you prefer email and/or RT: to `bug-xml-invisible
at rt.cpan.org`, or through the web interface at
[http://rt.cpan.org/NoAuth/ReportBug.html?Queue=XML-Invisible](http://rt.cpan.org/NoAuth/ReportBug.html?Queue=XML-Invisible). I will be
notified, and then you'll automatically be notified of progress on your
bug as I make changes.

# LICENSE AND COPYRIGHT

Copyright 2018 Ed J.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

[http://www.perlfoundation.org/artistic\_license\_2\_0](http://www.perlfoundation.org/artistic_license_2_0)


Changes
- switch to use Test::Snapshot

0.06 2020-01-01
- stop putting element "type" in output
- make_parser can take a Pegex::Grammar object

0.05 2019-12-31
- make XML::LibXML be only suggested, not required

0.04 Fri 19 Oct 03:45:53 BST 2018
 - make flatten work right on single rule with +

0.03 Mon 24 Sep 06:03:10 BST 2018
 - attributes hash not array

0.02 Mon 24 Sep 05:46:54 BST 2018
 - autovivify children and attributes so no empty arrays

0.01 Sun 23 Sep 05:25:42 BST 2018
 - First version, released on an unsuspecting world.


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME             => 'XML::Invisible',
  AUTHOR           => q{Ed J <etj@cpan.org>},
  VERSION_FROM     => 'lib/XML/Invisible.pm',
  ABSTRACT_FROM    => 'lib/XML/Invisible.pm',
  LICENSE          => 'artistic_2',
  MIN_PERL_VERSION => '5.008001',
  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => '6.64', # TEST_REQUIRES
  },
  TEST_REQUIRES => {
    'Test::More' => '0.88', # done_testing
    'Import::Into' => '0',
  },
  PREREQ_PM => {
    'Pegex' => '0.64',
    'XML::LibXML' => '0',
  },
  dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
  clean => { FILES => 'XML-Invisible-*' },
  META_MERGE => {
    "meta-spec" => { version => 2 },
    dynamic_config => 0,
    resources => {
      x_IRC => 'irc://irc.perl.org/#graphql-perl',
      repository => {
        type => 'git',
        url => 'git@github.com:mohawk2/xml-invisible.git',
        web => 'https://github.com/mohawk2/xml-invisible',
      },
      bugtracker  => {
        web => 'https://github.com/mohawk2/xml-invisible/issues',
      },
      license => [ 'http://dev.perl.org/licenses/' ],
    },
    prereqs => {
      develop => {
        requires => {
          'Test::Pod::Coverage' => '1.08',
          'Test::Pod' => '1.22',
          'Pod::Markdown' => 0,
        },
      },
    },
  },
);

sub MY::postamble {
  <<EOF;
pure_all :: README.md

README.md : \$(VERSION_FROM)
\tpod2markdown \$< >\$\@
EOF
}


lib/XML/Invisible/Receiver.pm
package XML::Invisible::Receiver;

use strict;
use warnings;
use base 'Pegex::Receiver';

sub gotrule {
  my $self = shift;
  my $param = (@_ == 1) ? $_[0] : [ @_ ];
  return unless defined $param;
  my $parser = $self->{parser};
  my $rule = $parser->{rule};
  my $parentrule = $parser->{parent};
  my ($attr, $flatten) = @{$parentrule}{qw(-wrap -flat)};
  return $param if $flatten;
  $param = [ $param ] if ref $param ne 'ARRAY';
  $param = $self->flatten($param);
  my %ret = (nodename => $rule, type => ($attr ? 'attr' : 'element'));
  for (@$param) {
    if (!ref $_ or !$_->{type}) {
      push @{ $ret{children} }, $_;
    } elsif ($_->{type} eq 'attr') {
      $ret{attributes}{$_->{nodename}} = join '', _get_values($_);
    } else {
      die "Unknown entity type '$_->{type}'";
    }
  }
  delete $ret{type} if $ret{type} eq 'element';
  \%ret;
}

sub _get_values {
  my ($node) = @_;
  map ref($_) ? _get_values($_) : $_, @{ $node->{children} };
}

1;

__END__
=head1 NAME

XML::Invisible::Receiver - XML::Invisible Pegex AST constructor

=head1 SYNOPSIS

  my $grammar = Pegex::Grammar->new(text => $grammar_text);
  my $parser = Pegex::Parser->new(
    grammar => $grammar,
    receiver => XML::Invisible::Receiver->new,
  );
  my $got = $parser->parse($ixml_text);

=head1 DESCRIPTION

Subclass of L<Pegex::Receiver> to turn Pegex parsing events into data
usable by L<XML::Invisible>.

The AST returned represents an XML document with a hash-ref with these keys:

=over

=item nodename

=item attributes

A hash-ref mapping name to value, since XML attributes can each only
occur once on an element, and the ordering is semantically meaningless.

=item children

An array-ref of child nodes. If the node is a simple scalar, it is a
text node.

=back

=cut


lib/XML/Invisible.pm
package XML::Invisible;

use strict;
use warnings;
use Exporter qw(import);
use Pegex::Grammar;
use Pegex::Parser;
use XML::Invisible::Receiver;

our $VERSION = '0.06';
our @EXPORT_OK = qw(make_parser ast2xml make_canonicaliser);

use constant DEBUG => $ENV{XML_INVISIBLE_DEBUG};

sub make_parser {
  my ($grammar) = @_;
  $grammar = Pegex::Grammar->new(text => $grammar) if !ref $grammar;
  my $parser = Pegex::Parser->new(
    grammar => $grammar,
    receiver => XML::Invisible::Receiver->new,
    debug => DEBUG,
  );
  sub {
    my ($ixml_text) = @_;
    $parser->parse($ixml_text);
  };
}

sub make_canonicaliser {
  my ($grammar_text) = @_;
  require Pegex::Compiler;
  require Pegex::Grammar::Atoms;
  my $grammar_tree = Pegex::Compiler->new->parse($grammar_text)->tree;
  my $toprule = $grammar_tree->{'+toprule'};
  my $atoms = _atoms2canonical(Pegex::Grammar::Atoms->atoms);
  sub {
    my ($ast) = @_;
    my @results = _extract_canonical($atoms, $ast, $grammar_tree, $toprule);
    return undef if grep !defined, @results;
    join '', @results;
  };
}

my %ATOM2SPECIAL = (
  ALL => "",
  BLANK => " ",
  BREAK => "\n",
  BS => "\x08",
  CONTROL => "\x00",
  CR => "\r",
  DOS => "\r\n",
  EOL => "\n",
  EOS => "",
  FF => "\x0C",
  HICHAR => "\x7f",
  NL => "\n",
  TAB => "\t",
  WORD => "a",
  WS => " ",
  _ => "",
  __ => " ",
  ws => "",
  ws1 => "",
  ws2 => " ",
);
sub _atoms2canonical {
  my ($atoms) = @_;
  my %lookup;
  for my $atom (keys %$atoms) {
    my $c = $atoms->{$atom};
    if (exists $ATOM2SPECIAL{$atom}) {
      $c = $ATOM2SPECIAL{$atom};
    } elsif ($c =~ s/^\\//) {
      # all good
    } elsif ($c =~ s/^\[//) {
      $c = substr $c, 0, 1;
    }
    $lookup{$atom} = $c;
  }
  \%lookup;
}

sub _extract_canonical {
  my ($atoms, $elt, $grammar_tree, $elt_sought, $grammar_frag) = @_;
  $grammar_frag ||= $grammar_tree->{$elt_sought};
  if (defined($elt)) {
    return undef if !ref $elt; # just text node - not valid
    return undef if defined($elt_sought) and $elt_sought ne $elt->{nodename};
  }
  if ($grammar_frag->{'.rgx'}) {
    # RE, so parent of text nodes
    return $elt->{children} ? join('', @{$elt->{children}}) : $elt->{nodename};
  }
  if (my $all = $grammar_frag->{'.all'}) {
    # sequence of productions
    my ($childcount, @results) = (0);
    for my $i (0..$#$all) {
      my $child = $elt->{children}[$childcount];
      my $all_frag = $all->[$i];
      if ($all_frag->{'-skip'}) {
        $child = undef;
      } else {
        $childcount++;
      }
      my @partial = _extract_canonical(
        $atoms, $child, $grammar_tree, undef, $all_frag,
      );
      return undef if grep !defined, @partial; # any non-match
      push @results, @partial;
    }
    return @results;
  } elsif (my $ref = $grammar_frag->{'.ref'}) {
    return $atoms->{$ref} if exists $atoms->{$ref};
    return undef if defined($elt) and $elt->{nodename} ne $ref;
    my $new_frag = $grammar_tree->{$ref};
    if ($new_frag->{'.ref'}) {
      # this one is just a single-child empty if it's a match
      return undef if defined($elt) and @{$elt->{children}} != 1;
      my $child = defined($elt) ? $elt->{children}[0] : undef;
      return _extract_canonical(
        $atoms, $child, $grammar_tree, $new_frag->{'.ref'}, $new_frag,
      );
    }
    # treat ourselves as if we're the ref-ed to thing
    return _extract_canonical(
      $atoms, $elt, $grammar_tree, $ref, $new_frag,
    );
  }
}

my $xml_loaded = 0;
sub ast2xml {
  do { require XML::LibXML; $xml_loaded = 1 } unless $xml_loaded;
  my ($ast) = @_;
  my $doc = XML::LibXML->createDocument("1.0", "UTF-8");
  $doc->addChild(_item2elt($ast));
  $doc;
}

sub _item2elt {
  my ($item) = @_;
  die "Unknown item '$item' passed" if ref $item ne 'HASH';
  my $elt = XML::LibXML::Element->new($item->{nodename});
  my $attrs = $item->{attributes} || {};
  $elt->setAttribute($_, $attrs->{$_}) for keys %$attrs;
  for (@{ $item->{children} || [] }) {
    if (!ref) {
      $elt->appendTextNode($_);
    } else {
      $elt->addChild(_item2elt($_));
    }
  }
  $elt;
}

1;

__END__
=head1 NAME

XML::Invisible - transform "invisible XML" documents into XML using a grammar

=begin markdown

# PROJECT STATUS

| OS      |  Build status |
|:-------:|--------------:|
| Linux   | [![Build Status](https://travis-ci.org/mohawk2/xml-invisible.svg?branch=master)](https://travis-ci.org/mohawk2/xml-invisible) |

[![CPAN version](https://badge.fury.io/pl/XML-Invisible.svg)](https://metacpan.org/pod/XML::Invisible) [![Coverage Status](https://coveralls.io/repos/github/mohawk2/xml-invisible/badge.svg?branch=master)](https://coveralls.io/github/mohawk2/xml-invisible?branch=master)

=end markdown

=head1 SYNOPSIS

  use XML::Invisible qw(make_parser);
  my $transformer = make_parser(from_file($ixmlspec));
  my $ast = $transformer->(from_file($ixml_input));

  use XML::Invisible qw(make_parser ast2xml);
  my $transformer = make_parser(from_file($ixmlspec));
  my $xmldoc = ast2xml($transformer->(from_file($ixml_input)));
  to_file($outputfile, $xmldoc->toStringC14N(1));

  # or, with conventional pre-compiled Pegex grammar
  my $parser = Pegex::Parser->new(
    grammar => My::Thing::Grammar->new,
    receiver => XML::Invisible::Receiver->new,
  );
  my $got = $parser->parse($input);

  # from command line
  cpanm XML::Invisible XML::Twig
  perl -MXML::Invisible=make_parser,ast2xml -e \
    'print ast2xml(make_parser(join "", <>)->("(a+b)"))->toStringC14N(1)' \
    examples/arith-grammar.ixml | xml_pp

  # canonicalise a document
  use XML::Invisible qw(make_parser make_canonicaliser);
  my $ixml_grammar = from_file('examples/arith-grammar.ixml');
  my $transformer = make_parser($ixml_grammar);
  my $ast = $transformer->(from_file($ixml_input));
  my $canonicaliser = make_canonicaliser($ixml_grammar);
  my $canonical = $canonicaliser->($ast);

=head1 DESCRIPTION

An implementation of Steven Pemberton's Invisible XML concept, using
L<Pegex>. Supply it with your grammar, in Pegex format (slightly
different from Steven's specification due to differences between his
notation and Pegex's), it returns you a function, which you can call to
transform "invisible XML" documents into actual XML.

This is largely a Pegex "receiver" class that exploits the C<+> and C<->
syntax in rules in slightly unintended ways, and a wrapper to make
this operate.

=head1 GRAMMAR SYNTAX

See L<Pegex::Syntax>. Generally, all rules will result in an XML
element. All terminals will need to capture with C<()> (see example
below).

However, if you specify a dependent token with C<+> it will
instead become an attribute (equivalent of Steven's C<@>). If C<->,
this will "flatten" (equivalent of Steven's C<->) - the children will
be included without making an element of that node. Since in Pegex any
element can be skipped entirely with C<.>, you can use that instead of
C<-> to omit terminals.

E.g.

  expr: +open -arith +close
  open: /( LPAREN )/
  close: /( RPAREN )/
  arith: left -op right
  left: +name
  right: -name
  name: /(a)/ | /(b)/
  op: +sign
  sign: /( PLUS )/

When given C<(a+b)> yields:

  <expr open="(" sign="+" close=")">
    <left name="a"/>
    <right>b</right>
  </expr>

=head1 FUNCTIONS

=head2 make_parser

Exportable. Returns a function that when called with an "invisible XML"
document, it will return an abstract syntax tree (AST), of the general form:

  {
    nodename => 'expr',
    attributes => { open => '(', sign => '+', close => ')' },
    children => [
      { nodename => 'left', attributes => { name => 'a' } },
      { nodename => 'right', children => [ 'b' ] },
    ],
  }

Arguments:

=over

=item an "invisible XML" Pegex grammar specification, OR a L<Pegex::Grammar> object

=back

See L<XML::Invisible::Receiver> for more.

=head2 ast2xml

Exportable. Turns an AST, as output by L</make_parser>,
from L<XML::Invisible::Receiver> into an object of class
L<XML::LibXML::Document>. Needs L<XML::LibXML> installed, which as of
version 0.05 of this module is only a suggested dependency, not required.

Arguments:

=over

=item an AST from L<XML::Invisible::Receiver>

=back

=head2 make_canonicaliser

Exportable. Returns a function that when called with an AST as produced
from a document by a L</make_parser>, returns a canonical version of
the original document, or C<undef> if it failed.

Arguments:

=over

=item an XML::Invisible grammar

=back

It uses a few heuristics:

=over

=item literals that are 0-1 (C<?>) or any number (C<*>) will be omitted

=item literals that are at least one (C<+>) will be inserted once

=item literal whitespace will be treated specially

=back

=head1 DEBUGGING

To debug, set environment variable C<XML_INVISIBLE_DEBUG> to a true value.

=head1 SEE ALSO

L<Pegex>

L<https://homepages.cwi.nl/~steven/ixml/> - Steven Pemberton's Invisible XML page

=head1 AUTHOR

Ed J, C<< <etj at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests on
L<https://github.com/mohawk2/xml-invisible/issues>.

Or, if you prefer email and/or RT: to C<bug-xml-invisible
at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=XML-Invisible>. I will be
notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 LICENSE AND COPYRIGHT

Copyright 2018 Ed J.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

=cut
